official/vision/beta/projects/yolo/losses/yolo_loss.py

```diff
@@ -0,0 +1,654 @@
+        20% extended radius around each pixel that this specific pixel can 
+        predict values for a center at. the center can range from 0 - value/2 
+        to 1 + value/2, this value is set in the yolo filter, and resused here. 
+        there should be one value for scale_xy for each level from min_level to 
+        max_level. One for each FPN path.
+      max_deltas: `Dict[float]` for gradient clipping to apply to the box loss 
+        for each FPN path.
+      label_smoothing: `Dict[float]` for how much to smooth the loss on the 
+        classes for each FPN path.
+      use_scaled_loss: `bool` for whether to use the scaled loss 
+        or the traditional loss.
+      update_on_repeat: `bool` for whether to replace with the newest or 
+        the best value when an index is consumed by multiple objects. 
+    """
+    if use_scaled_loss:
+      loss_type = "scaled"
+    else:
+      loss_type = "darknet"
+
+    self._loss_dict = {}
+    for key in keys:
+      self._loss_dict[key] = LOSSES[loss_type](
+          classes=classes,
+          anchors=anchors,
+          mask=masks[key],
+          truth_thresh=truth_thresholds[key],
+          ignore_thresh=ignore_thresholds[key],
+          loss_type=loss_types[key],
+          iou_normalizer=iou_normalizers[key],
+          cls_normalizer=cls_normalizers[key],
+          obj_normalizer=obj_normalizers[key],
+          box_type=box_types[key],
+          objectness_smooth=objectness_smooths[key],
+          max_delta=max_deltas[key],
+          path_stride=path_strides[key],
+          scale_x_y=scale_xys[key],
+          update_on_repeat=update_on_repeat,
+          label_smoothing=label_smoothing)
+
+  def __call__(self, ground_truth, predictions, use_reduced_logs=True):
+    metric_dict = defaultdict(dict)
+    metric_dict['net']['box'] = 0
+    metric_dict['net']['class'] = 0
+    metric_dict['net']['conf'] = 0
+
+    loss_val, metric_loss = 0, 0
+    num_replicas_in_sync = tf.distribute.get_strategy().num_replicas_in_sync
+
+    for key in predictions.keys():
+      (_loss, _loss_box, _loss_conf, _loss_class, _mean_loss, _avg_iou,
+       _avg_obj) = self._loss_dict[key](ground_truth['true_conf'][key],
+                                        ground_truth['inds'][key],
+                                        ground_truth['upds'][key],
+                                        ground_truth['bbox'],
+                                        ground_truth['classes'],
+                                        predictions[key])
+
+      # after computing the loss, scale loss as needed for aggregation
+      # across FPN levels
+      _loss = self._loss_dict[key].post_path_aggregation(
+          _loss, ground_truth, predictions)
+          
+      # after completing the scaling of the loss on each replica, handle
+      # scaling the loss for mergeing the loss across replicas
+      _loss = self._loss_dict[key].cross_replica_aggregation(
```

Comment:
Is the whole purpose of cross_replica_aggregation to scale the loss? if so, maybe the scaling factor can be provided as an argument at creation time. (i.e. loss_class(**kwargs, loss_scale: float = 1.0)).